server/render/index.js

```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
 
 	next();
 }
+
+/**
+ * A generic middleware that identifies if the current request fits the conditions for being SSRable.
+ *
+ * Warning: Having serverSideRender=true is not sufficient. The app-level logic needs be applied before truly SSRing.
+ *
+ * @param {object}   context  The entire request context
+ * @param {function} next     As all middlewares, will call next in the sequence
+ */
+export function setShouldServerSideRender( context, next ) {
+	context.serverSideRender = Object.keys( context.query ).length === 0; // 0 query parameters expected
+
+	next();
+}
+
+/**
+ * Applies all the app-related checks for server side rendering.
+ *
+ * Warning: If this returns true, it is not sufficient for the page to be SSRed. Returning true from here is a
+ * pre-condition for SSR and the result needs to be corroborated with context.serverSideRender
+ * (context.serverSideRender is set to boolean by the middlewares, depending, in general, on the query params).
+ *
+ * Warning: if you think about calling this method or adding these conditions to the middlewares themselves (the ones
+ * that set context.serverSideRender), think twice: the context may not have all the necessary values;
+ * examples: context.layout, context.user may be set by other middlewares).
```

Comment:
Maybe also add a note that section specific criteria should be handled by section level middleware, see https://github.com/Automattic/wp-calypso/pull/25554/files#r198843610